prow/config/branch_protection.go

```diff
@@ -43,3 +49,100 @@ type Branch struct {
 	Contexts []string `json:"require-contexts,omitempty"`
 	Pushers  []string `json:"allow-push,omitempty"`
 }
+
+func (bp BranchProtection) isSet() bool {
+	switch {
+	case bp.ProtectTested:
+		return true
+	case bp.Protect != nil:
+		return true
+	case len(bp.Orgs) > 0:
+		return true
+	case len(bp.Pushers) > 0:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *Config) GetBranchProtection(org, repo, branch string) (*Branch, error) {
+	if !c.BranchProtection.isSet() {
+		return nil, nil
+	}
+
+	var protect *bool
+	pushers := sets.NewString()
+	contexts := sets.NewString()
+
+	update := func(b *bool, c, p []string) {
+		if b != nil {
+			protect = b
+
+		}
+		pushers.Insert(p...)
+		contexts.Insert(c...)
+	}
+
+	if c.BranchProtection.ProtectTested {
+		// Adding ProwJobs
+		prowContexts := prowJobRequirements(org, repo, branch, c.Presubmits)
+		if len(prowContexts) > 0 {
+			yes := true
+			update(&yes, prowContexts, nil)
+		}
+	}
+
+	update(c.BranchProtection.Protect, c.BranchProtection.Contexts, c.BranchProtection.Pushers)
+	if orgP, exists := c.BranchProtection.Orgs[org]; exists {
+		update(orgP.Protect, orgP.Contexts, orgP.Pushers)
+		if repoP, exists := orgP.Repos[repo]; exists {
+			update(repoP.Protect, repoP.Contexts, repoP.Pushers)
+			if branchP, exists := repoP.Branches[branch]; exists {
+				update(branchP.Protect, branchP.Contexts, branchP.Pushers)
+			}
+		}
+	}
+
+	if protect == nil {
+		return nil, fmt.Errorf("protect should not be nil")
+	}
+
+	if contexts.Len() > 0 || pushers.Len() > 0 {
+		if !*protect {
+			return nil, fmt.Errorf("setting pushers or contexts requires protection")
+		}
+	}
+
+	return &Branch{
+		Protect:  protect,
+		Contexts: contexts.List(),
+		Pushers:  pushers.List(),
+	}, nil
+}
+
+func jobRequirements(jobs []Presubmit, branch string, after bool) []string {
```

Comment:
Can you do one commit that is simple code movement and another with additions/deletions/edits? hard to tell what to review and how

Reply:
that s not practical however I move the test changes in another commit and restored the recursive way of finding orgs, repo and branches to decrease the amount of code changes.

Reply:
I think it would be helpful to rebase away all of the unnecessary commits here. Hard to understand what is going on with 18 commits.

Reply:
Sorry, I'm totally wrong... that's the other PR. The commits here are good. Just keep the original prow/cmd/branchprotector test cases?